### PR TITLE
ON HOLD: Double spending prevention using Secp256k1 keys on TrustChain

### PR DESCRIPTION
**Pull request by xoriole**
Double spending prevention using Secp256k1 signatures on transactions.

### CODE REVIEW

**Comment by qstokkink** [comment]
Ok to test

**Comment by xoriole** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

File: ipv8/attestation/trustchain/block.py [hunk 1]
```diff
@@ -273,6 +290,14 @@ def err(reason):
                 linklinked = database.get_linked(link)
                 if linklinked is not None and linklinked.hash != self.hash:
                     err("Double countersign fraud")
+                    if 'double_sig' in self.transaction and self.transaction['double_sig'] \
+                        and 'double_sig' in linklinked.transaction and linklinked.transactino['double_sig']:
```
Comment (qstokkink) [inline ipv8/attestation/trustchain/block.py]: `transactino`?

File: ipv8/keyvault/doublesign.py [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+            return self.zero()
+        else:
+            l = (p.y - q.y) / (p.x - q.x)
+
+        x = l ** 2 - (p.x + q.x)
+        y = l * (p.x - x) - p.y
+        return self.point(x, y)
+
+    def sub(self, lhs, rhs):
+        return lhs + -rhs
```
Comment (qstokkink) [inline ipv8/keyvault/doublesign.py]: `+ -` is redundant

File: test/util.py [hunk 3]
```diff
@@ -166,4 +166,4 @@ def twisted_wrapper(arg):
     """
     if isinstance(arg, (int, long)):
         return lambda x: deferred(arg)(inlineCallbacks(x))
-    return deferred(timeout=1)(inlineCallbacks(arg))
+    return deferred(timeout=10)(inlineCallbacks(arg))
```
Comment (qstokkink) [inline test/util.py]: Please change this per test, not globally.

File: ipv8/keyvault/doublesign.py (hunk 2 above)
Comment (xoriole) [reply ipv8/keyvault/doublesign.py]: This is the way it should be. Point negation and addition.

**Comment by xoriole** [comment]
retest this please

**Comment by xoriole** [comment]
retest this please

**Comment by xoriole** [comment]
retest this please

**Comment by qstokkink** [comment]
`test/keyvault/test_doublesign.py` needs to be added to https://github.com/Tribler/py-ipv8/blob/master/test_classes_list.txt

File: test/base.py [hunk 4]
```diff
@@ -95,7 +95,7 @@ def add_node_to_experiment(self, node):
         self.nodes.append(node)
 
     @inlineCallbacks
-    def deliver_messages(self, timeout=.1):
+    def deliver_messages(self, timeout=10):
```
Comment (qstokkink) [inline test/base.py]: Please don't change the base timeouts.

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

File: ipv8/attestation/trustchain/block.py [hunk 5]
```diff
@@ -409,3 +441,10 @@ def invalid():
         The block violates at least one validation rule
         """
         pass
+
+    @staticmethod
+    def double_spend():
+        """
+        The block violates at least one validation rule
```
Comment (devos50) [inline ipv8/attestation/trustchain/block.py]: Does the docstring match with the method name?

File: ipv8/attestation/trustchain/community.py [hunk 6]
```diff
@@ -51,11 +51,13 @@ class TrustChainCommunity(Community):
     def __init__(self, *args, **kwargs):
         working_directory = kwargs.pop('working_directory', '')
         db_name = kwargs.pop('db_name', self.DB_NAME)
+        double_sign = kwargs.pop('double_sign', False)
```
Comment (devos50) [inline ipv8/attestation/trustchain/community.py]: Please rename this to `check_for_double_sign` or something similar 👍

File: ipv8/keyvault/crypto.py [hunk 7]
```diff
@@ -132,3 +136,44 @@ def is_valid_signature(self, ec, data, signature):
             return ec.verify(signature, data)
         except:
             return False
+
+    def is_valid_double_signature(self, data, signature):
+        """
+        Returns True if signature is valid
```
Comment (devos50) [inline ipv8/keyvault/crypto.py]: Minor detail: please be consistent with ending docstrings with a period or not :)

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
Merged into the `nuclear_backdoor` branch.

**Pull request closed without merging**